firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +60,142 @@
   /** The event listener for the query that raises events asynchronously. */
   private final ExecutorEventListener<ViewSnapshot> asyncEventListener;
 
+  static class CallbackList {
+    void run() {
+      for (Runnable callback : callbacks) {
+        if (callback != null) {
+          callback.run();
+        }
+      }
+    }
+
+    void add(Runnable callback) {
+      callbacks.add(callback);
+    }
+
+    private final List<Runnable> callbacks = new ArrayList<>();
+  }
+
+  public static class StopListenerSupportFragment extends android.support.v4.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @SuppressWarnings("deprecation")
+  public static class StopListenerFragment extends android.app.Fragment {
+    @Override
+    public void onStop() {
+      super.onStop();
+      callbacks.run();
+    }
+
+    CallbackList callbacks = new CallbackList();
+  }
+
+  @Nullable
+  private static <T> T castFragment(Class<T> fragmentClass, @Nullable Object fragment, String tag) {
+    try {
+      return fragmentClass.cast(fragment);
+    } catch (ClassCastException e) {
+      throw new IllegalStateException(
+          "Fragment with tag '"
+              + tag
+              + "' is a "
+              + fragment.getClass().getName()
+              + " but should be a "
+              + fragmentClass.getName());
+    }
+  }
+
+  private static final String SUPPORT_FRAGMENT_TAG = "FirestoreOnStopObserverSupportFragment";
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private static WeakHashMap<Activity, WeakReference<StopListenerSupportFragment>>
+      supportFragmentByActivity = new WeakHashMap<>();
+  private static WeakHashMap<Activity, WeakReference<StopListenerFragment>> fragmentByActivity =
+      new WeakHashMap<>();
+
+  /**
+   * Implementation for non-FragmentActivity Activity's. Unfortunatly, all Fragment related
+   * classes/methods with nonFragmentActivityActivity's are deprecated, imply that almost everything
+   * in this function is deprecated.
+   */
+  @SuppressWarnings("deprecation")
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    hardAssert(
+        !(activity instanceof FragmentActivity),
+        "onActivityStopCallOnce must be called with a *non*-FragmentActivity Activity.");
+
+    synchronized (activity) {
```

Comment:
Synchronizing on a user-supplied object is (rarely) a good idea so we should avoid that on principle. Additionally, multiple concurrent calls on separate activities doesn't actually protect our map.

I think this should synchronize on the map object not the activity.